pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	statsRequestTimeout     = time.Second * 2
+	readyStalenessThreshold = time.Minute * 5
```

Comment:
I really don't like the staleness check (and definitely not one that's expressed in minutes), since it will result in cluster-wide outages when proxy loses connectivity to the control plane.

Also, in cases when connection wasn't actively closed by the other side, you won't see the disconnect at the TCP level for hours, so this isn't the most reliable indicator anyway.